### DISCUSSION TITLE
ICON 2.0 code size limit

### DISCUSSION THREAD

**Discussion by danielvo11** [icon-project/btp#39] [Q&A] [answered]
As I hear from Dongdwan Kim, ICON 2.0 code size limit is 512kb. 

But as I calculated from current cost configuration in Gicon testnet and the rule define in https://www.icondev.io/docs/transaction-fees.

Max Step Invoke Limit: 2'500'000'000 steps

Step cost: 

- contractCreate: 1'000'000'000 steps
- contractSet: 30'000 steps
- C: 100'000

```
cumulativeStepUsed = max( ∑ βi Si + C, C ) < 2.5 billions steps
```

In case of deploy contract:

```
Step = S(contractCreate)*β(contractCreate) + S(contractSet)*β(contractSet) + C
= 1*1'000'000'000 + 30'000*ContractSize + 100'000 < 2.5 billions steps
```

=> ContractSize < 48 Kbytes

Current JAVA BMV contract size is 128kb, I have to reduce `contractSet` cost parameter in my local node to be able to deploy. 

I don't know how can I get the number of 512kb contract size limit. Will ICON 2.0 config with different cost parameters?

/cc @flowersinthesand @quocle108

**Comment by cow-hs** [comment, marked as answer]
I am curious about the source for "As I hear from Dongdwan Kim, ICON 2.0 code size limit is 512kb".
Because as described "=> ContractSize < 48 kbytes" is currently correct.
If you want to increase it further, you need to adjust the value (MaxStepLimit or contractSet) according to the internal governance policy.

Unfortunally, There are no immediate plans to increase that size by changing the policy.
Therefore, if it is unavoidable, please separate the contract by logic and distribute it separately.

**Reply by flowersinthesand** [reply]
> I am curious about the source for "As I hear from Dongdwan Kim, ICON 2.0 code size limit is 512kb".

Oh I'm told that from Donghyun when asking limitation in ICON 2.0.

BTW we are not sure how to separate smart contracts that decode events now due to its complexity. We will find a way to do that and share it probably today.

**Reply by cow-hs** [reply]
There was a misunderstanding in communication.

It was decided to support up to 64kb according to the ICON policy.
However, since the value of 64kb is not allowed by the network in operation right away, it is determined through the agreement of governance, it is recommended to consider this and write a contract.

**Reply by quocle108** [reply]
@leeheonseung 

Here is our current situation:

The total code size of BMV now about 60kb. We've optimized the code, instead of storing all event's structures to decode the event data, we store the size of event's structures for unnecessary events.

We also have separated to 2 smart contracts: verify receipt hash (36kb), decode the events (24kb).

> It was decided to support up to 64kb according to the ICON policy.
> However, since the value of 64kb is not allowed by the network in operation right away, it is determined through the agreement of governance, it is recommended to consider this and write a contract.

Will this policy apply soon? if so do we need separate the contract? 
_"Therefore, if it is unavoidable, please separate the contract by logic and distribute it separately."_

**Reply by cow-hs** [reply]
@quocle108
Not coming soon. :(
We have agreed to keep 64kb only, but there is still a lot to be discussed about when and how to apply this.
We will inform you when it is decided.

**Reply by flowersinthesand** [reply]
We concluded to split BMV on ICON. It won't affect the schedule.
